### PR TITLE
PEP 484 Type Hints for pyUmbral

### PR DESCRIPTION
**Pull request by KPrasch**
#### Goal
Passing mypy checks, PEP 484 compliance.

##### Adds type annotations to the following modules:
- [X] `_pre.py`
- [X] `config.py`
- [ ] `curvebn.py`
- [X] `dem.py`
- [x] `fragments.py`
- [X] `keys.py`
- [x] `opensll.py` - Skipped
- [X] `params.py`
- [ ] `point.py`
- [X] `pre.py`
- [X] `signing.py`
- [X] `utils.py`

See: #183

### CODE REVIEW

**Comment by KPrasch** [comment]
The remaining modules (curvebn, openssl, and point) use ctypes that are a little harder to annotate.  Perhaps we can introduce types with > 1 Pull Request.

**Comment by tuxxy** [comment]
> The remaining modules (curvebn, openssl, and point) use ctypes that are a little harder to annotate.

Yes, I think it's just `CData`. Given the generic name, I have tried to name the params to be similar to the C structs that are passed in. IE: `EC_GROUP`.

**Comment by KPrasch** [comment]
We dont have 100% coverage or accuracy - but mypy errors have been vastly reduced.  

I'd like to take this PR off WIP while we work on annotating the remaining parts of the codebase in a future PR.

File: umbral/pre.py [hunk 1]
```diff
@@ -1,45 +1,47 @@
...
-                 point_noninteractive=None,
-                 delegating_pubkey: UmbralPublicKey = None,
-                 receiving_pubkey: UmbralPublicKey = None,
-                 verifying_pubkey=None):
+                 point_e: Optional[Point] = None,
+                 point_v: Optional[Union[int, Point]] = None,
+                 bn_sig: Optional[Union[int, CurveBN]] = None,
+                 point_e_prime: Optional[Point] = None,
+                 point_v_prime: Optional[Union[int, Point]] = None,
+                 point_noninteractive: Optional[Union[int, Point]] = None,
```
Comment (cygnusv) [inline umbral/pre.py]: All `Points` above are just `Optional[Point]`, not `Optional[Union[int, Point]]`. Same for `bn_sig`, but with `CurveBN`.

File: umbral/pre.py [hunk 2]
```diff
@@ -177,7 +180,8 @@ def verify(self) -> bool:
         s = self._bn_sig
         h = CurveBN.hash(e, v, params=self._umbral_params)
 
-        return s * g == v + (h * e)
+        result = s * g == v + (h * e)      # type: bool
```
Comment (cygnusv) [inline umbral/pre.py]: Isn't this redundant? We already said above that the result is `bool`.

File: umbral/pre.py [hunk 3]
```diff
@@ -260,15 +264,16 @@ def __eq__(self, other):
             # Again, it's hard to imagine why.
             return False
 
-    def __hash__(self):
+    @typing.no_type_check
```
Comment (cygnusv) [inline umbral/pre.py]: Just curious, why?

File: umbral/signing.py [hunk 4]
```diff
@@ -81,22 +82,22 @@ def __len__(self):
     def __add__(self, other):
```
Comment (cygnusv) [inline umbral/signing.py]: Why `__add__` and `__len__` don't have type hints, but `__bytes__`, `__radd__', and `__eq__` do?

File: umbral/utils.py [hunk 5]
```diff
@@ -1,12 +1,16 @@
...
 from cryptography.hazmat.backends import default_backend
 from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives.asymmetric.ec import SECP192R1, SECP256K1, SECP256R1, SECP384R1
 from cryptography.hazmat.primitives.kdf.hkdf import HKDF
-from cryptography.hazmat.primitives.asymmetric import ec
+
 from umbral import openssl
 
-def lambda_coeff(id_i, selected_ids):
+CurveTypes = Union[Type[SECP256K1], Type[SECP192R1], Type[SECP384R1], Type[SECP256R1]]
```
Comment (cygnusv) [inline umbral/utils.py]: We can get rid of `SECP192R1`.

**Comment by KPrasch** [comment]
Cool - Ill peck away at the changes.

 Feel free to add more comments  - I combed through the changeset - but much of the annotations were auto-generated with monkeytype so more eyes on the code is a good thing.

File: umbral/signing.py (hunk 4 above)
Comment (KPrasch) [reply umbral/signing.py]: other than `__eq__`, They just haven't been covered yet.  As for `__eq__` an implementation change may be needed as `__eq__` needs to be able to take any `object`.

File: umbral/utils.py (hunk 5 above)
Comment (KPrasch) [reply umbral/utils.py]: Well, it's a type that is passed in during test runtime, an it needs to be included in the annotation for it to pass type checks.  Perhaps we need curves type-aliases

File: umbral/pre.py (hunk 3 above)
Comment (KPrasch) [reply umbral/pre.py]: there is a tricky check inside the genexp here.

File: umbral/pre.py (hunk 2 above)
Comment (KPrasch) [reply umbral/pre.py]: Yeah - perhaps redundant - Looking for precision on the charts, the return value type and the return type of the arithmetic operators are different.

File: umbral/fragments.py [hunk 6]
```diff
@@ -69,19 +69,19 @@ def to_bytes(self):
         return self._id + key + ni + commitment + xcoord + signature
 
     def verify(self,
-               signing_pubkey,
+               signing_pubkey: UmbralPublicKey,
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply umbral/fragments.py]: Perhaps a type alias, skip,  or `Any` will suffice

File: umbral/config.py [hunk 7]
```diff
@@ -18,19 +22,19 @@ def __set_curve_by_default(cls):
...
     @classmethod
-    def params(cls):
+    def params(cls) -> UmbralParameters:
         if not cls.__params:
             cls.__set_curve_by_default()
         return cls.__params
 
     @classmethod
-    def curve(cls):
+    def curve(cls) -> Union[Type[SECP256R1], Type[SECP256K1]]:
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply umbral/config.py]: Yeah - Sounds like a good Idea - we need a single sourced alias for curves, IMO

File: umbral/_pre.py [hunk 8]
```diff
@@ -110,11 +111,11 @@ def assess_cfrag_correctness(cfrag, capsule: "Capsule"):
            & correct_rk_commitment
 
 
-def verify_kfrag(kfrag,
+def verify_kfrag(kfrag: 'KFrag',
                  delegating_pubkey: UmbralPublicKey,
-                 signing_pubkey,
+                 signing_pubkey: UmbralPublicKey,
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply umbral/_pre.py]: Hmm, what shall we do in the meantime?

**Comment by KPrasch** [comment]
What do you all think about addressing the request for changes in a future PR?

File: umbral/pre.py (hunk 1 above)
Comment (KPrasch) [reply umbral/pre.py]: Yeah - @jMyles and were are a bit stumped on how int is passed here during test-runtime.

File: umbral/_pre.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply umbral/_pre.py]: Yep - looks like we need to implement a custom subprtocol: https://mypy.readthedocs.io/en/latest/protocols.html#simple-user-defined-protocols

File: umbral/pre.py (hunk 2 above)
Comment (cygnusv) [reply umbral/pre.py]: Different? I don't see why they're different.

**Pull request merged**